image/cas/interface.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cas implements generic content-addressable storage.
+package cas
+
+import (
+	"io"
+)
+
+// Engine represents a content-addressable storage engine.
+type Engine interface {
+
+	// Put adds a new blob to the store.
+	Put(writer io.Reader) (digest string, err error)
```

Comment:
The interface doesn't need to be nearly as heavy weight, but I'd recommend taking inspiration from the [distribution interfaces](https://godoc.org/github.com/docker/distribution#BlobWriter). They support a number of uses that aren't obvious and there are few mistakes made there, as well:

1. It's much easier to control the copy if you have something to write to, rather than passing in a writer and hoping for the best.
2. Consider the case of re-entrant writes. I don't think they are necessary here, but having a repeatable identifier for resuming a failed write exists there.
3. Consider the transaction boundary at the point this write completes. This is interesting when two writers are trying to write the same blob. This must ensure that only one succeeds but they may race each other.
4. Consider the use of handles, rather than directly accessing the readers and writers. This will allow you to have better control over the scope of a reader and writer. 
5. Be very careful with `ReadFrom`, if you rely on it for performance. The last point helps here but this becomes a nightmare if you wrap interfaces.
6. Consider splitting the front end interface from the driver. Make the front end concrete. There are many details that are important for the driver that aren't important for the user of the interface. (We made this mistake in distribution).

All in all, this looks fine, but I hope these points can help.

Camlistore also is a great resource for structuring these kinds of interfaces, as well.